### PR TITLE
Add 4 New Array Functions

### PR DESCRIPTION
**Pull request by jokoho48**
Add 4 New Array Functions to Find First Entry of TypeOf, TypeName, Null or Nil in Array

### CODE REVIEW

**Comment by Killswitch00** [comment]
Since the "find" script command finds the **first** array element or string-within-string, these function names could be shortened somewhat. Also, "empty" in BIS scripting parlance is "nil". How about calling them

 * `CBA_fnc_findNil` finds the first _nil_ entry
 * `CBA_fnc_findNull` finds the first _null_ entry
 * `CBA_fnc_findTypeOf` finds the first entry of a certain type.

**Comment by MikeMatrix** [comment]
I like @Killswitch00's names a lot more. They're very similar, to the point and short to type.

**Comment by jokoho48** [comment]
is everything fine?

**Comment by Killswitch00** [comment]
The unit tests for the three new functios are missing and the added unit test file could be renamed to, say, "test_find.sqf".

**Comment by jokoho48** [comment]
my question is did i need a test function? or not

File: addons/arrays/test_find.sqf [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+TEST_OP(str _original,==,str _expected,_fn);
+TEST_OP(str _result,==,str _expected,_fn);
+
+
+
+
+_fn = "CBA_fnc_findTypOf";
+LOG("Testing " + _fn);
+
+TEST_DEFINED("CBA_fnc_findTypOf","");
```
Comment (Killswitch00) [inline addons/arrays/test_find.sqf]: I can see that this unit test file hasn't been tested at all. Run it using
```
[] call compile preprocessFileLineNumbers "\x\cba\addons\arrays\test_find.sqf";
```

File: addons/arrays/test_find.sqf [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+TEST_DEFINED("CBA_fnc_findNil","");
+
+// Find First Nil Entry
+_original = ["", player, "", player, nil, "", player, nil];
+_result = ["", player, "", player, nil, "", player, nil] call CBA_fnc_findNil;
+_expected = 4;
+TEST_OP(str _original,==,str _expected,_fn);
+TEST_OP(str _result,==,str _expected,_fn);
+
+
```
Comment (Killswitch00) [inline addons/arrays/test_find.sqf]: The three tests below all do the same thing as the first test above - look for the first literal `nil` element in the test arrays. Keep the first one, and change the three redundant tests to use
 * an array with a variable in it somewhere that is nil
 * an array with a code element in it that returns nil
 * an empty input array ([])

File: addons/arrays/test_find.sqf [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+_original = ["", player, "", player, "", "", player, displayNull];
+_result = ["", player, "", player, "", "", player, displayNull] call CBA_fnc_findNull;
+_expected = 7;
+TEST_OP(str _original,==,str _expected,_fn);
+TEST_OP(str _result,==,str _expected,_fn);
+
```
Comment (Killswitch00) [inline addons/arrays/test_find.sqf]: Good - these four tests above test for different kinds of null values (but a test with controlNull is missing). It also needs tests for
* an input array with a variable in it somewhere that is null-valued (for example objNull).
* an input array with a code element in it that returns one of the null values.
* an input array that is empty.

File: addons/arrays/test_findNil.sqf [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+// ----------------------------------------------------------------------------
+
+private ["_expected", "_result", "_fn"];
+
+_fn = "CBA_fnc_findNil";
+LOG("Testing " + _fn);
+
+TEST_DEFINED("CBA_fnc_findNil","");
+
+// Use of embeded nil
```
Comment (Killswitch00) [inline addons/arrays/test_findNil.sqf]: "embedded"

File: addons/arrays/test_findNull.sqf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+LOG("Testing " + _fn);
+
+TEST_DEFINED("CBA_fnc_findNull","");
+
+// Use of embedded null
+_result = ["", objNull, player, 3] call CBA_fnc_findNull;
+_expected = 1;
+TEST_OP(_result,==,_expected,_fn);
+
+// Only find first Null
```
Comment (Killswitch00) [inline addons/arrays/test_findNull.sqf]: "null", not "Null"

File: addons/arrays/fnc_findTypeOf.sqf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+---------------------------------------------------------------------------- */
+#include "script_component.hpp"
+
+scopeName "main";
+
+params [["_array", [], [[]]], ["_typeOf", nil, [objNull, ""]]];
+
+if (isNil "_typeOf" || {_array isEqualTo []}) exitWith {-1};
+
+if (IS_OBJECT(_typeOf) then {
```
Comment (Killswitch00) [inline addons/arrays/fnc_findTypeOf.sqf]: I see that this hasn't been tested in-game at all.

File: addons/arrays/fnc_findTypeName.sqf [hunk 7]
```diff
@@ -0,0 +1,48 @@
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_findTypeName
+
+Description:
+    A function that returns the index of the first entry of the given type in an array.
+
+Parameters:
+    0: Array
+    1: TypeName, if parameter is a string, that contains a case insensitive typename, it will be used. Otherwies typename of the variable will be used.
```
Comment (Killswitch00) [inline addons/arrays/fnc_findTypeName.sqf]: "Otherwise"

Comment (MikeMatrix) [inline addons/arrays/fnc_findTypeName.sqf]: God dammit :D

File: addons/arrays/test_findTypeOf.sqf [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+_result = [["", player, 5], player] call CBA_fnc_findTypeOf;
+_expected = 1;
+TEST_OP(_result,==,_expected,_fn);
+
+// Find first only
+_result = [["", player, 5, objNull, player], typeOf player] call CBA_fnc_findTypeOf;
+_expected = 1;
+TEST_OP(_result,==,_expected,_fn);
+
+// Finf first only, mixed solutions in array
```
Comment (Killswitch00) [inline addons/arrays/test_findTypeOf.sqf]: "Finf" --> "Find"

File: addons/arrays/test_findTypeOf.sqf [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+_result = [["", typeOf player, 5, objNull, player], typeOf player] call CBA_fnc_findTypeOf;
+_expected = 1;
+TEST_OP(_result,==,_expected,_fn);
+
+// Return not found, when no matching elemnt is in array
```
Comment (Killswitch00) [inline addons/arrays/test_findTypeOf.sqf]: "element"

File: addons/arrays/fnc_findTypeName.sqf [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+#define TYPENAMES ["ARRAY", "BOOL", "CODE", "CONFIG", "CONTROL", "DISPLAY", "GROUP", "LOCATION", "OBJECT", "SCALAR", "SCRIPT", "SIDE", "STRING", "TASK", "TEXT", "TEAM_MEMBER", "NAMESPACE"]
+
+scopeName "main";
+
+params [["_array", [], [[]]], "_typeName"];
+
+if (isNil "_typeName" || {_array isEqualTo []}) exitWith {-1};
+
+// If a string is given, tansform to uppercase for type matching
```
Comment (Killswitch00) [inline addons/arrays/fnc_findTypeName.sqf]: "transform"

Comment (ViperMaul) [inline addons/arrays/fnc_findTypeName.sqf]: A better word for "transform" in this context is to use "convert"

**Pull request merged**